### PR TITLE
[BUGFIX] Fixed systemtic testing error as a result of raising Exception() and BaseException()

### PR DESCRIPTION
**Pull request by cartertinney**
* Refactored all tests that need arbitrary, non-specific excpetions to use `arbitrary_exception` or `arbitrary_base_exception` top level fixtures instead
* Fixed any broken tests discovered along the way

### CODE REVIEW

File: azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py [hunk 1]
```diff
@@ -151,9 +151,10 @@ def fail_set_auth_provider(self, op):
             pipeline_stages_base.PipelineRootStage,
             "_execute_op",
             side_effect=fail_set_auth_provider,
+            autospec=True,
         )
 
-        with pytest.raises(fake_exception.__class__):
+        with pytest.raises(unexpected_exception.__class__):
             ProvisioningPipeline(input_security_client)
```
Comment (BertKleewein) [inline azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py]: add assert e_info.value?

File: azure-iot-device/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+that exception handling was hiding other errors (also caught by an "except: Exception" block).
+
+The solution is to use a subclass of Exception or BaseException that is not defined anywhere else,
+thus guaranteeing that it will be unexpected and unhandled except by broad all-encompassing
+handling. Furthermore, because the exception in question is derived from either Exception or
+BaseException, but is not itself an instance of either, tests checking that the exception in
+question is raised will not spuriously pass due to different exceptions being raised.
+
+For consistency, and to prevent confusion, please do this ONLY by using one of the follwing
+fxitures.
```
Comment (olivakar) [inline azure-iot-device/tests/conftest.py]: >fxitures [](start = 0, length = 8)

spell

File: azure-iot-device/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+You may (and should!) still use exceptions defined elsewhere for specific, non-arbitrary exceptions
+(e.g. testing specific exceptions)
+"""
+
+
+@pytest.fixture
```
Comment (olivakar) [inline azure-iot-device/tests/conftest.py]: DO you want to use scope ?

File: azure-iot-device/tests/iothub/pipeline/test_pipeline_stages_iothub.py [hunk 4]
```diff
@@ -322,29 +335,31 @@ def test_update_sas_token_operation(self, stage):
...
-    @pytest.mark.it(
-        "Handles any Exceptions raised by the UpdateSasTokenOperation and passes them into the unhandled exception handler"
-    )
-    def test_raises_exception(self, stage, mocker):
-        threading.current_thread().name = "not_pipeline"
+    # @pytest.mark.it(
+    #     "Handles any Exceptions raised by the UpdateSasTokenOperation and passes them into the unhandled exception handler"
+    # )
+    # def test_raises_exception(self, stage, mocker, unhandled_error_handler, arbitrary_exception):
+    #     threading.current_thread().name = "not_pipeline"
```
Comment (olivakar) [inline azure-iot-device/tests/iothub/pipeline/test_pipeline_stages_iothub.py]: will these be commented ?

Comment (cartertinney) [reply azure-iot-device/tests/iothub/pipeline/test_pipeline_stages_iothub.py]: Forgot these, will fix now

---
In reply to: [332195314](https://github.com/Azure/azure-iot-sdk-python/pull/330#discussion_r332195314) [](ancestors = 332195314)

**Pull request merged**